network/p2p/acp118/aggregator.go

```diff
@@ -0,0 +1,294 @@
+
+	totalStakeWeight := uint64(0)
+	nodeIDsToValidator := make(map[ids.NodeID]indexedValidator)
+	for i, v := range validators {
+		totalStakeWeight, err = math.Add[uint64](totalStakeWeight, v.Weight)
+		if err != nil {
+			return nil, err
+		}
+
+		// Check that the validator set provided is valid
+		if _, ok := nodeIDsToValidator[v.NodeID]; ok {
+			return nil, fmt.Errorf("%w: %s", ErrDuplicateValidator, v.NodeID)
+		}
+
+		nodeIDsToValidator[v.NodeID] = indexedValidator{
+			Index:     i,
+			Validator: v,
+		}
+	}
+
+	minThreshold := (totalStakeWeight * quorumNumMin) / quorumDenMin
+	maxThreshold := (totalStakeWeight * quorumNumMax) / quorumDenMax
+
+	job := aggregationJob{
+		log:                s.log,
+		client:             s.client,
+		requestBytes:       requestBytes,
+		messageBytes:       message.Bytes(),
+		minThreshold:       minThreshold,
+		maxThreshold:       maxThreshold,
+		validators:         validators,
+		nodeIDsToValidator: nodeIDsToValidator,
+		totalStakeWeight:   totalStakeWeight,
+		signatures:         make([]*bls.Signature, 0),
+		signerBitSet:       set.NewBits(),
+		sem:                sem,
+		done:               done,
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			// Try to return whatever progress we have made if the context is
+			// cancelled early
+			signature, err := job.GetResult()
+			if err != nil {
+				return nil, err
+			}
+
+			return warp.NewMessage(message, signature)
+		case result := <-done:
+			if result.Err != nil {
+				return nil, result.Err
+			}
+
+			return warp.NewMessage(message, result.Signature)
+		case sem <- struct{}{}:
+			if err := job.SendRequest(ctx); err != nil {
+				return nil, err
+			}
+		}
+	}
+}
+
+type aggregationJob struct {
+	log                logging.Logger
+	client             *p2p.Client
+	requestBytes       []byte
+	messageBytes       []byte
+	minThreshold       uint64
+	maxThreshold       uint64
+	validators         []Validator
+	index              int
+	nodeIDsToValidator map[ids.NodeID]indexedValidator
+	totalStakeWeight   uint64
+
+	lock                  sync.Mutex
+	aggregatedStakeWeight uint64
+	failedStakeWeight     uint64
+	signatures            []*bls.Signature
+	signerBitSet          set.Bits
+
+	sem  <-chan struct{}
+	done chan result
+}
+
+func (a *aggregationJob) GetResult() (*warp.BitSetSignature, error) {
+	a.lock.Lock()
+	defer a.lock.Unlock()
+
+	return a.getResult()
+}
+
+func (a *aggregationJob) getResult() (*warp.BitSetSignature, error) {
+	aggregateSignature, err := bls.AggregateSignatures(a.signatures)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", ErrFailedAggregation, err)
+	}
+
+	bitSetSignature := &warp.BitSetSignature{
+		Signers:   a.signerBitSet.Bytes(),
+		Signature: [bls.SignatureLen]byte{},
+	}
+
+	copy(bitSetSignature.Signature[:], bls.SignatureToBytes(aggregateSignature))
+	return bitSetSignature, nil
+}
+
+func (a *aggregationJob) SendRequest(ctx context.Context) error {
+	if len(a.validators) == 0 {
```

Comment:
Should we add a check that `index < len(a.validators)` as well? It looks like it's possible to call `SendRequest` after all validators have been queried in the case where `s.maxPending == len(a.validators)`

Reply:
I refactored some code around so this edge-case doesn't exist anymore